parse.py

```diff
@@ -0,0 +1,222 @@
+from functions import common
+from functions import exponents_and_logs
+from functions import statistic
+from functions import trigonometry
+import exceptions.exceptions as exceptions
+
+from pyparsing import (
+    Literal,
+    Word,
+    Group,
+    Forward,
+    alphas,
+    alphanums,
+    Regex,
+    ParseException,
+    CaselessKeyword,
+    Suppress,
+    delimitedList,
+)
+
+
+class Parser(object):
+    """A class used to parse mathematical expressions.
+
+    Methods:
+        evaluate(expression, expected=None, parse_all=True)
+            Returns the result of an expression in string form.
+    """
+    _symbol_stack = []
+
+    def push_first(self: object, tokens: list) -> None:
+        """Push first element of 'tokens' onto symbol stack.
+        
+        Used properly during parsing, this function can turn infix notation into postfix notation.
+    
+        Args:
+            tokens (list): Tokens from which first token must be taken.
+        """
+        self._symbol_stack.append(tokens[0])
+
+    def push_unary_operator(self: object, tokens: list) -> None:
+        """Processes unary operators applied to atom and appends appropriate operators to symbol stack.
+        
+        For now, we have only two possible unary operators ('-' and '+'). 
+        +' does nothing and '-' turns the atom negative in odd numbers.
+    
+        Args:
+            tokens (list): Tokens belonging to atom. Unary operators to apply to atom will be at start of list.
+        """
+        # We will track parity of unary '-'s to push the 'unary -' symbol onto the stack only once per atom and improve runtime.
+        is_negative = False
+        for e in tokens:
+            if e == '-':
+                is_negative = not is_negative
```

Comment:
is_negative =True isn't correct here???

Reply:
I don't understand this comment. The term is positive by default hence is_negative = false. Negative signs makes negative numbers positive and positive numbers negative hence the is_negative = not is_negative.

Reply:
ah i get it thanks